bundle/Command/TagContentByTypesCommand.php

```diff
@@ -0,0 +1,198 @@
+
+        $this->style->newLine();
+        $this->style->progressStart($totalResults);
+
+        $this->repository->beginTransaction();
+
+        for ($offset = 0; $offset < $totalResults; $offset += $batchSize) {
+            $query->offset = $offset;
+            $query->limit = $batchSize;
+            $searchResults = $this->repository->getSearchService()->findContent($query);
+
+            if ($this->input->getOption('field-identifiers') === null) {
+                $fieldIdentifiers = $this->configResolver->getParameter('tag_command_default_field_identifiers', 'ngsite');
+            } else {
+                $fieldIdentifiers = $this->getFieldIdentifiers();
+            }
+
+            $this->repository->beginTransaction();
+
+            foreach ($searchResults->searchHits as $searchHit) {
+                $result = $this->repository->sudo(
+                    function () use ($searchHit, $fieldIdentifiers): ?int {
+                        /** @var \Ibexa\Contracts\Core\Repository\Values\Content\Content $content */
+                        $content = $searchHit->valueObject;
+
+                        $contentDraft = $this->repository->getContentService()->createContentDraft($content->contentInfo);
+                        $contentUpdateStruct = $this->repository->getContentService()->newContentUpdateStruct();
+
+                        foreach ($fieldIdentifiers as $fieldIdentifier) {
+                            if (!$this->hasField($content, $fieldIdentifier)) {
+                                continue;
+                            }
+
+                            if (!$content->getField($fieldIdentifier)->value instanceof TagFieldValue) {
+                                $this->style->error(sprintf('Field with identifier %s must be a type of eztags', $fieldIdentifier));
+                                return Command::FAILURE;
+                            }
+
+                            $alreadyAssignedTags = $content->getFieldValue($fieldIdentifier)->tags;
+                            $tagsToAssign = array_filter($alreadyAssignedTags, fn ($alreadyAssignedTag) => $this->getTag()->id !== $alreadyAssignedTag->id);
+                            $tagsToAssign[] = $this->getTag();
+                            $contentUpdateStruct->setField($fieldIdentifier, new TagFieldValue($tagsToAssign));
+
+                            break;
+                        }
+
+                        $this->repository->getContentService()->updateContent($contentDraft->versionInfo, $contentUpdateStruct);
+                        $this->repository->getContentService()->publishVersion($contentDraft->versionInfo);
+
+                        $this->style->progressAdvance();
+
+                        return null;
+                    }
+                );
+
+                if ($result === Command::FAILURE) {
+                    return Command::FAILURE;
+                }
+            }
+        }
+
+        $this->repository->commit();
```

Comment:
You are missing error handling - if there are any exceptions, you need to rollback the transaction.

Reply:
solved in 872ee9fa